filters/include/pcl/filters/experimental/grid_filter_base.h

```diff
@@ -0,0 +1,214 @@
+/*
+ * SPDX-License-Identifier: BSD-3-Clause
+ *
+ *  Point Cloud Library (PCL) - www.pointclouds.org
+ *  Copyright (c) 2020-, Open Perception
+ *
+ *  All rights reserved
+ */
+
+#pragma once
+
+#include <pcl/filters/experimental/transform_filter.h>
+
+#include <limits>
+
+namespace pcl {
+namespace experimental {
+
+#define GET_POINT_TYPE(GridStructT) typename GridStructT::PointCloud::PointType
+
+template <typename GridStruct, typename PointT = GET_POINT_TYPE(GridStruct)>
+class GridFilterBase : public TransformFilter<GridStruct, PointT> {
+protected:
+  using PointCloud = typename TransformFilter<GridStruct, PointT>::PointCloud;
+  using PointCloudPtr = typename PointCloud::Ptr;
+  using PointCloudConstPtr = typename PointCloud::ConstPtr;
+
+  /**
+   * \brief Filter point clouds based on a templated grid structure
+   * \details Used as the base class for grid based filters, e.g. VoxelGrid,
+   * ApproximateVoxelGrid. For grid based filters with different behavior, one can
+   * implement a custom grid structure and pass to this class. The templated grid
+   * structure should have two main components: member functions for filtering
+   * opterations, Grid member attribute for storing the information of the smaller
+   * spaces divided by the grid (e.g. voxels in VoxelGrid).
+   *
+   * Requirements of the grid structure:
+   *  1. Three member functions (setUp, addPointToGrid, filterGrid) the grid structure
+   * are called in applyFilter and thus required to declare.
+   *  2. A Grid member attribute is required, it can be any type with built-in iterator
+   * and size() function, e.g. STL container or custom grid with iterator
+   * \ingroup filters
+   */
+public:
+  /** \brief Empty constructor. */
+  GridFilterBase()
+  : downsample_all_data_(true)
+  , save_leaf_layout_(false)
+  , filter_field_name_("")
+  , filter_limit_min_(std::numeric_limits<double>::min())
+  , filter_limit_max_(std::numeric_limits<double>::max())
+  , filter_limit_negative_(false)
+  , min_points_per_voxel_(0)
+  {}
```

Comment:
Can we use `downsample_all_data_ = true;` style initialization instead?

Reply:
Maybe use field initializers?

Reply:
For consistence with other filters, I changed to use field initializer for `filter_name_`. (in another PR) 
Other default values are now initialized directly